homeassistant/components/livisi/coordinator.py

```diff
@@ -70,67 +63,66 @@ async def async_setup(self) -> None:
                 livisi_connection_data=livisi_connection_data
             )
-        controller_info = await self.aiolivisi.async_get_controller()
-
-        controller_data = controller_info
+        controller_data = await self.aiolivisi.async_get_controller()
         if controller_data.get("controllerType") == "Avatar":
             self.port = AVATAR_PORT
             self.is_avatar = True
         else:
             self.port = CLASSIC_PORT
             self.is_avatar = False
-        self.serial_number = controller_data.get("serialNumber")
-        self.controller_type = controller_data.get("controllerType")
-        self.os_version = controller_data.get("osVersion")
+        self.serial_number = controller_data["serialNumber"]
+        self.controller_type = controller_data["controllerType"]
+        self.os_version = controller_data["osVersion"]
 
-    async def async_get_devices(self) -> list:
+    async def async_get_devices(self) -> list[dict[str, Any]]:
         """Set the discovered devices list."""
-        shc_devices = await self.aiolivisi.async_get_devices()
-        if bool(shc_devices) is False:
-            shc_devices = await self.aiolivisi.async_get_devices()
-        return shc_devices
+        return await self.aiolivisi.async_get_devices()
 
-    async def async_get_pss_state(self, capability):
+    async def async_get_pss_state(self, capability: str) -> bool:
         """Set the PSS state."""
-        response = await self.aiolivisi.async_get_pss_state(capability[1:])
+        response: dict[str, Any] = await self.aiolivisi.async_get_pss_state(
+            capability[1:]
+        )
         if response is None:
-            return
-        on_state = response.get("onState")
-        return on_state.get("value")
+            return None
+        on_state = response["onState"]
+        return on_state["value"]
 
-    async def async_set_all_rooms(self):
+    async def async_set_all_rooms(self) -> None:
         """Set the room list."""
-        response = await self.aiolivisi.async_get_all_rooms()
+        response: list[dict[str, Any]] = await self.aiolivisi.async_get_all_rooms()
 
         for available_room in response:
-            self.rooms[available_room.get("id")] = available_room.get("config").get(
-                "name"
-            )
+            available_room_config: dict[str, Any] = available_room["config"]
+            self.rooms[available_room["id"]] = available_room_config["name"]
 
     def on_data(self, event_data: LivisiEvent) -> None:
         """Define a handler to fire when the data is received."""
-        source = event_data.source
         if event_data.onState is not None:
             device_id_state: dict = {
-                "id": source,
+                "id": event_data.source,
                 "state": event_data.onState,
             }
-            self.async_set_updated_data(device_id_state)
+            dispatcher_send(self.hass, LIVISI_STATE_CHANGE, device_id_state)
```

Comment:
Will this run in the asyncio event loop or in another thread? Make sure to use our async api (`async_dispatcher_send`) in async context and our sync api (`dispatcher_send`) in sync context.

Reply:
This runs in the asyncio event loop. I modify the code to use `async_dispatcher_send` .